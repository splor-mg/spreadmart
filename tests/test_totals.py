import duckdb
import pandas as pd
import pytest
from deepdiff import DeepDiff

def query(sql):
    con = duckdb.connect('data/db.duckdb')
    cur = con.execute(sql)

    columns = [col[0] for col in cur.description]

    result = []
    for row in cur.fetchall():
        result.append(dict(zip(columns, row)))

    con.close()
    
    return result
    
def read(path):
    result = pd.read_csv(path).to_dict('records')
    return result

def test_totals():
    expected = read('tests/data/test_totals_expected.csv')
    
    sql = """select ano_particao as ANO, round(sum(vr_empenhado), 2) as VL_EMP
             from ft_despesa 
             where ano_particao between 2002 and 2022 
             group by ano_particao
             order by ano_particao
           """
    result = query(sql)

    diff = DeepDiff(expected, result, verbose_level=2)
    
    if diff:
        pytest.fail(diff.pretty())
