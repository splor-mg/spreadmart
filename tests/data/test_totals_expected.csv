ANO,VL_EMP
2002,19359441437.9
2003,20263459410.28
2004,21726608097.05
2005,25292069834.95
2006,29018789168.21
2007,32443148300.45
2008,39299362850.28
2009,40263460915.06
2010,46051174100.08
2011,54693566620.55
2012,63172171470.52
2013,71906529761.8
2014,75512924847.61
2015,85119036085.58
2016,88129098217.78
2017,98391669163.23
2018,102986641548.54
2019,108180584581.56
2020,107107557290.39
2021,128891804747.7
2022,116489779176.87
